src/librawspeed/common/Spline.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+    RawSpeed - RAW file decoder.
+
+    Copyright (C) 2017 Robert Bieber
+
+    This library is free software; you can redistribute it and/or
+    modify it under the terms of the GNU Lesser General Public
+    License as published by the Free Software Foundation; either
+    version 2 of the License, or (at your option) any later version.
+
+    This library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+    Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public
+    License along with this library; if not, write to the Free Software
+    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+*/
+
+#include "common/Spline.h"
+#include "common/Common.h" // for ushort16
+#include "common/Point.h"  // for iPoint2D
+#include <vector>          // for vector
+
+namespace rawspeed {
+
+// This is an implementation of the cubic spline algorithm found on
+// Wikipedia at https://en.wikipedia.org/wiki/Spline_(mathematics)
+std::vector<ushort16>
+Spline::calculateCurve(std::vector<iPoint2D> control_points) {
```

Comment:
Just so that i have asked: the dcraw's `cubic_spline()` is indeed the cubic spline, just like this code, not counting bugs?
I.e. not counting bugs, they are supposed to produce the same curve?

Reply:
Mathematically I believe they should be doing the same thing, but as far as I can tell (I haven't attempted to fully comprehend dcraw's implementation) we're using different algorithms to accomplish it.  So it's entirely possible that we'll end up infinitesimally different coefficients due to rounding

Reply:
ok, so at most rounding issues.

Reply:
I believe so, but if it will make you feel better I could try comparing the curve values between my and dcraw's implementation for the same image to see if there's any difference

Reply:
As long as both the implementations are of the *same* algorithm, and the IIQ de-quadrant-ization is actually meant to be corrected via that algorithm, i suppose it is ok. We don't want to mimic some other implementation. What it needs is tests against trustworthy spline implementation in some mathematical software, but that is for the future.

Reply:
Okay, we should be good then.  It's definitely doing the same fundamental thing